File: RaftSwift/Sources/RaftGRPCTransport/Services/PeerService.swift
import GRPCCore
import RaftCore

struct PeerService: Raft_RaftPeer.SimpleServiceProtocol {
    let node: any RaftNodeProtocol

    init(node: any RaftNodeProtocol) {
        self.node = node
    }

    func requestVote(request: Raft_RequestVoteRequest, context: ServerContext) async throws -> Raft_RequestVoteResponse {
        let response = await node.requestVote(request: RequestVoteRequest(
            term: Int(request.term),
            candidateID: Int(request.candidateID),
            lastLogIndex: Int(request.lastLogIndex),
            lastLogTerm: Int(request.lastLogTerm),
        ))

        return .with { grpcResponse in
            grpcResponse.term = UInt64(response.term)
            grpcResponse.voteGranted = response.voteGranted
        }
    }

    func appendEntries(request: Raft_AppendEntriesRequest, context: ServerContext) async throws -> Raft_AppendEntriesResponse {
        let response = await node.appendEntries(request: AppendEntriesRequest(
            term: Int(request.term),
            leaderID: Int(request.leaderID),
            prevLogIndex: Int(request.prevLogIndex),
            prevLogTerm: Int(request.prevLogTerm),
            entries: request.entries.map { LogEntry(term: Int($0.term), key: $0.key, value: $0.value) },
            leaderCommit: Int(request.leaderCommit),
        ))

        return .with { grpcResponse in
            grpcResponse.term = UInt64(response.term)
            grpcResponse.success = response.success
        }
    }

    func installSnapshot(request: Raft_InstallSnapshotRequest, context: ServerContext) async throws -> Raft_InstallSnapshotResponse {
        let response = try await node.installSnapshot(request: InstallSnapshotRequest(
            term: Int(request.term),
            leaderID: Int(request.leaderID),
            snapshot: Snapshot(
                lastIncludedIndex: Int(request.snapshot.lastIncludedIndex),
                lastIncludedTerm: Int(request.snapshot.lastIncludedTerm),
                stateMachine: request.snapshot.stateMachine,
            ),
        ))

        return .with { grpcResponse in
            grpcResponse.term = UInt64(response.term)
        }
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/transport/grpc/services/PeerService.kt
package raft.transport.grpc.services

import raft.Peer
import raft.RaftPeerGrpcKt
import raft.appendEntriesResponse
import raft.core.node.RaftNode
import raft.core.utils.peer.AppendEntriesRequest
import raft.core.utils.peer.InstallSnapshotRequest
import raft.core.utils.peer.RequestVoteRequest
import raft.core.utils.types.LogEntry
import raft.core.utils.types.Snapshot
import raft.installSnapshotResponse
import raft.requestVoteResponse
import raft.transport.grpc.utils.extensions.toGRPC

/**
 * gRPC service implementation for Raft peer communication.
 */
class PeerService(
    private val node: RaftNode
) : RaftPeerGrpcKt.RaftPeerCoroutineImplBase() {

    override suspend fun requestVote(request: Peer.RequestVoteRequest): Peer.RequestVoteResponse {
        val response = node.requestVote(
            request = RequestVoteRequest(
                term = request.term.toInt(),
                candidateId = request.candidateId,
                lastLogIndex = request.lastLogIndex.toInt(),
                lastLogTerm = request.lastLogTerm.toInt()
            )
        )

        return requestVoteResponse {
            term = response.term.toLong()
            voteGranted = response.voteGranted
        }
    }

    override suspend fun appendEntries(request: Peer.AppendEntriesRequest): Peer.AppendEntriesResponse {
        val response = node.appendEntries(
            request = AppendEntriesRequest(
                term = request.term.toInt(),
                leaderId = request.leaderId,
                prevLogIndex = request.prevLogIndex.toInt(),
                prevLogTerm = request.prevLogTerm.toInt(),
                entries = request.entriesList.map { entry ->
                    LogEntry(
                        term = entry.term.toInt(),
                        key = entry.key,
                        value = entry.value
                    )
                },
                leaderCommit = request.leaderCommit.toInt()
            )
        )

        return appendEntriesResponse {
            term = response.term.toLong()
            success = response.success
        }
    }

    override suspend fun installSnapshot(request: Peer.InstallSnapshotRequest): Peer.InstallSnapshotResponse {
        val response = node.installSnapshot(
            request = InstallSnapshotRequest(
                term = request.term.toInt(),
                leaderId = request.leaderId,
                snapshot = Snapshot(
                    lastIncludedIndex = request.snapshot.lastIncludedIndex.toInt(),
                    lastIncludedTerm = request.snapshot.lastIncludedTerm.toInt(),
                    stateMachine = request.snapshot.stateMachineMap
                )
            )
        )

        return installSnapshotResponse {
            term = response.term.toLong()
        }
    }
}